ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,205 @@
+  Status RemoveAll(std::function<void(ValueType)> callback_destroy_item);
+
+ private:
+  std::unordered_map<KeyType, ValueType> m_ng_items_map;
+  std::deque<KeyType> m_lru;
+  int m_depth;
+  absl::Mutex m_mutex;
+
+  // Test class
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_SameKeyMultiThread_Test;
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_RemoveItemTest_Test;
+};
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::NgraphDataCache(int depth)
+    : m_depth(depth) {}
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::~NgraphDataCache() {
+  m_ng_items_map.clear();
+  m_lru.clear();
+}
+
+template <typename KeyType, typename ValueType>
+Status NgraphDataCache<KeyType, ValueType>::RemoveItem(
+    KeyType key, std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  if (m_ng_items_map.find(key) != m_ng_items_map.end()) {
+    callback_destroy_item(m_ng_items_map.at(key));
+    m_ng_items_map.erase(key);
+    m_lru.pop_back();
+  }
+  if (m_ng_items_map.size() != m_lru.size()) {
+    return errors::Internal(
+        "Error occured: size of m_ng_items_map is not same as that of m_lru");
+  }
+  return Status::OK();
+}
+
+template <typename KeyType, typename ValueType>
+Status NgraphDataCache<KeyType, ValueType>::RemoveItem(KeyType key) {
+  return RemoveItem(key, [](ValueType) {});
+}
+
+template <typename KeyType, typename ValueType>
+Status NgraphDataCache<KeyType, ValueType>::RemoveAll(
+    std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  for (auto it = m_ng_items_map.begin(); it != m_ng_items_map.end(); it++) {
+    callback_destroy_item(it->second);
+  }
+  if (m_ng_items_map.size() != m_lru.size()) {
+    return errors::Internal(
+        "Error occured: size of m_ng_items_map is not same as that of m_lru");
+  }
+  m_ng_items_map.erase(m_ng_items_map.begin(), m_ng_items_map.end());
+  m_lru.clear();
+  return Status::OK();
+}
+
+template <typename KeyType, typename ValueType>
+std::pair<Status, ValueType>
+NgraphDataCache<KeyType, ValueType>::LookUpOrCreate(
+    KeyType key,
+    std::function<std::pair<Status, ValueType>()> callback_create_item,
+    std::function<void(ValueType)> callback_destroy_item) {
+  bool found_in_cache;
+  // look up in the cache
+  {
+    absl::MutexLock lock(&m_mutex);
+    auto it = m_ng_items_map.find(key);
+    found_in_cache = (it != m_ng_items_map.end());
+    if (found_in_cache) {
+      return std::make_pair(Status::OK(), m_ng_items_map.at(key));
+    }
+  }
+  // Item not found in cache, create item
+  ValueType item;
+  pair<Status, ValueType> status_item_pair;
+  try {
+    status_item_pair = callback_create_item();
+  } catch (std::bad_function_call& exception) {
+    return std::make_pair(
+        errors::Internal(
```

Comment:
all the callbacks need to be try catched

Reply:
All other callbacks are calling destroy_item(),
not sure what is the right thing to do if destoy_item in invalid,
In case of LookUPOrCreate(), should I still continue to create the item or not,
and In case of RemoveItem(), should I still be destroying item from cache or not.